### PR TITLE
rpc: check nil blockmeta

### LINKED ISSUE #4319: RPC Panic on Nil Values
**Tendermint version**:

0.33.0

**ABCI app**:

Gaia

**What happened**:

RPC panics when calling `client.HTTP#Block(nil)`. Depending on the timing of this call and since `height` is `nil`, the actual height of the chain can be zero and thus `blockMeta` can be nil causing a nil panic.

```
runtime error: invalid memory address or nil pointer dereference" stack="goroutine 126 [running]:
runtime/debug.Stack(0xc00108b168, 0x4d4aa00, 0x5b8e090)
	/usr/local/go/src/runtime/debug/stack.go:24 +0x9d
github.com/tendermint/tendermint/rpc/lib/server.RecoverAndLogHandler.func1.1(0xc0010d4300, 0x52cb580, 0xc000fe8900, 0xbf80959394931700, 0x11a40003, 0x5baa360, 0xc0004b6600)
	/Users/aleksbez/go/pkg/mod/github.com/tendermint/tendermint@v0.33.0/rpc/lib/server/http_server.go:161 +0x486
panic(0x4d4aa00, 0x5b8e090)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/tendermint/tendermint/rpc/core.Block(0xc0010d4340, 0x0, 0x0, 0x0, 0x0)
	/Users/aleksbez/go/pkg/mod/github.com/tendermint/tendermint@v0.33.0/rpc/core/blocks.go:81 +0xd6
reflect.Value.call(0x4d39580, 0x50c2c08, 0x13, 0x4ec0793, 0x4, 0xc000f73b60, 0x2, 0x2, 0x2, 0x18, ...)
	/usr/local/go/src/reflect/value.go:460 +0x5f6
reflect.Value.Call(0x4d39580, 0x50c2c08, 0x13, 0xc000f73b60, 0x2, 0x2, 0x1, 0x1, 0x0)
	/usr/local/go/src/reflect/value.go:321 +0xb4
github.com/tendermint/tendermint/rpc/lib/server.makeJSONRPCHandler.func1(0x52c5380, 0xc0010d4300, 0xc0004b6600)
	/Users/aleksbez/go/pkg/mod/github.com/tendermint/tendermint@v0.33.0/rpc/lib/server/http_json_handler.go:103 +0x923
github.com/tendermint/tendermint/rpc/lib/server.handleInvalidJSONRPCPaths.func1(0x52c5380, 0xc0010d4300, 0xc0004b6600)
	/Users/aleksbez/go/pkg/mod/github.com/tendermint/tendermint@v0.33.0/rpc/lib/server/http_json_handler.go:127 +0x9c
net/http.HandlerFunc.ServeHTTP(0xc000fa9d10, 0x52c5380, 0xc0010d4300, 0xc0004b6600)
	/usr/local/go/src/net/http/server.go:2007 +0x44
net/http.(*ServeMux).ServeHTTP(0xc000fe6700, 0x52c5380, 0xc0010d4300, 0xc0004b6600)
	/usr/local/go/src/net/http/server.go:2387 +0x1bd
github.com/tendermint/tendermint/rpc/lib/server.maxBytesHandler.ServeHTTP(0x52a94c0, 0xc000fe6700, 0xf4240, 0x52c5380, 0xc0010d4300, 0xc0004b6600)
	/Users/aleksbez/go/pkg/mod/github.com/tendermint/tendermint@v0.33.0/rpc/lib/server/http_server.go:210 +0xcf
github.com/tendermint/tendermint/rpc/lib/server.RecoverAndLogHandler.func1(0x52c5a80, 0xc0010960e0, 0xc0004b6600)
	/Users/aleksbez/go/pkg/mod/github.com/tendermint/tendermint@v0.33.0/rpc/lib/server/http_server.go:183 +0x33f
net/http.HandlerFunc.ServeHTTP(0xc00033bef0, 0x52c5a80, 0xc0010960e0, 0xc0004b6600)
	/usr/local/go/src/net/http/server.go:2007 +0x44
net/http.serverHandler.ServeHTTP(0xc00102e000, 0x52c5a80, 0xc0010960e0, 0xc0004b6600)
	/usr/local/go/src/net/http/server.go:2802 +0xa4
net/http.(*conn).serve(0xc000512aa0, 0x52cab80, 0xc000085940)
	/usr/local/go/src/net/http/server.go:1890 +0x875
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:2928 +0x384
```

`blockMeta` should be checked if its nil before returning a result:

```go
func Block(ctx *rpctypes.Context, heightPtr *int64) (*ctypes.ResultBlock, error) {
	storeHeight := blockStore.Height()
	height, err := getHeight(storeHeight, heightPtr)
	if err != nil {
		return nil, err
	}

	blockMeta := blockStore.LoadBlockMeta(height) // blockMeta might be nil here but its not checked!!!
	block := blockStore.LoadBlock(height) // block might be nil here but thats OK
	return &ctypes.ResultBlock{BlockID: blockMeta.BlockID, Block: block}, nil
}
```

it should be something like the following:

```go
if blockMeta == nil {
	return &ctypes.ResultBlock{BlockID: types.BlockID{}, Block: block}, nil
}
return &ctypes.ResultBlock{BlockID: blockMeta.BlockID, Block: block}, nil
```

I would recommend going through the entire RPC core to make sure we check for `nil` values.

### PR DESCRIPTION
**Pull request by marbar3778**
- fixes #4319

- check if block meta is nil

Signed-off-by: Marko Baricevic <marbar3778@yahoo.com>

* [ ] Referenced an issue explaining the need for the change
* [ ] Updated all relevant documentation in docs
* [ ] Updated all code comments where relevant
* [ ] Wrote tests
* [ ] Updated CHANGELOG_PENDING.md

### CODE REVIEW

File: rpc/core/blocks.go [hunk 1]
```diff
@@ -76,8 +76,11 @@ func Block(ctx *rpctypes.Context, heightPtr *int64) (*ctypes.ResultBlock, error)
 		return nil, err
 	}
 
-	blockMeta := blockStore.LoadBlockMeta(height)
 	block := blockStore.LoadBlock(height)
```
Comment (melekes) [inline rpc/core/blocks.go]: so there's a block, but no block meta? how's this possible?

Comment (marbar3778) [reply rpc/core/blocks.go]: from my understanding this `return &ctypes.ResultBlock{BlockID: blockMeta.BlockID, Block: block}, nil` panics if there is no blockmeta to blockid

Comment (melekes) [reply rpc/core/blocks.go]: but why is there no block meta?

**Comment by marbar3778** [comment]
I don't believe there is block meta for height 0. 

Sent from Yahoo Mail on Android 
 
  On Sat, Jan 18, 2020 at 17:56, Anton Kaliaev<notifications@github.com> wrote:   
@melekes commented on this pull request.

In rpc/core/blocks.go:
> @@ -76,8 +76,11 @@ func Block(ctx *rpctypes.Context, heightPtr *int64) (*ctypes.ResultBlock, error)
 		return nil, err
 	}
 
-	blockMeta := blockStore.LoadBlockMeta(height)
 	block := blockStore.LoadBlock(height)

but why is there no block meta?

File: rpc/core/blocks.go (hunk 1 above)
Comment (melekes) [reply rpc/core/blocks.go]: > I don't believe there is block meta for height 0. 

Right. There's no block either (for height 0), but in `LoadBlock` we first load `BlockMeta` and, if it's `nil`, return `nil` block.

File: rpc/core/blocks.go [hunk 2]
```diff
@@ -88,6 +91,9 @@ func BlockByHash(ctx *rpctypes.Context, hash []byte) (*ctypes.ResultBlock, error
 	height := block.Height
```
Comment (melekes) [inline rpc/core/blocks.go]: this will panic too because block can be `nil`!

**Pull request merged**